### ISSUE TITLE
"griffiths" threshold measure should not be per-cluster

### ISSUE THREAD

**Issue by jonathangriffiths** [plger/scDblFinder#30]
Hi there,

Aaron pointed me towards this package and I think there's something I can help with. In the `doubletThresholding` function, the `"griffiths"` method is calculating deviations from the per-cluster medians. However, this is a problem when a cluster consists only of doublets, as cells may not deviate from the cluster average, which will be high. These clusters are quite common if you have very different cell types in your experiment.

Rather, in the paper I used this approach for, I calculated deviations per-sample. The idea being that different samples have different characteristics (timepoint, level of digestion etc.) and therefore should be handled separately. 

I thought about submitting a PR to tweak this, but I wasn't sure if you wanted the function to be used across samples (in which case `$cluster` can more or less be replaced with `$sample`), or on one sample at a time (in which case the code just gets very much simpler).

Cheers,
Jonny.

**Comment by plger** [comment]
Hi! thanks for bringing this up, wasn't sure how this should be implemented exactly and realize now that I misinterpreted the description in the paper.
The aim is just to offer users various possibilities (i.e. scDblFinder isn't using that method by default, but minimizes the misclassification rates and the deviation from expected doublet rates), so I'd definitely implement your way, but also 'degrade' this variant elegantly if 'sample' information isn't provided in the input. Feel free to do a PR to edit it if you want, otherwise I'll just make sure I got it right (here just for 1 sample):

1. you calculate the global median `M` and `MAD` of the score
2. for each cluster, you check the probability of its median # MADs from `M` based on a normal distribution with mean=`M` and and variance=`MAD`.
3. if the BH-corrected probability is <0.1, you consider the cluster to be doublets.

Is that correct?
(here, since this is about thresholding, we'd probably just then take a cell-level <0.1)

Thanks again,
Pierre-Luc

**Comment by jonathangriffiths** [comment]
Hi - sorry for the delay!

That is correct - where the scores were all per-cluster (technically per subcluster - because I wanted even more granular groups of cells in this particular dataset, as the main clusters were way too large). I found the doubletCells/scrublet scores to be a bit too variable to work on the per-cell level, often leaving things that were obviously doublets, so the clusters added a lot of robustness. A cell-level alternative might be to diffuse the scores between nearby cells, somehow.

I was planning to get to a PR, but it's a busy time right now - so if you're in a rush you might have to get to it yourself. I guess I'd get there early next week :)

Jonny.

**Comment by plger** [comment]
Thanks for the confirmation.
I made some basic changes, but still keeping it at the cell-level for the moment. I'll make a cluster-based alternative in addition when I have the time (no idea when).
(The smoothing of scores across the network is something that I already implemented prior to thresholding, so I'll just make sure this is available as an autonomous function that users can manually call before.)
